converters/google/convert.go

```diff
@@ -200,7 +205,34 @@ func (c *Converter) AddResource(r TerraformResource) error {
 			// The existance of a merge function signals that this tf resource maps to a
 			// patching operation on an API resource.
 			if mapper.merge != nil {
-				converted = mapper.merge(existing.converterAsset, converted)
+				converterAsset := existing.converterAsset
+
+				// The existence of a newIamUpdater function signals that this tf
+				// resource needs to be merged with existing remote IAM data to be useful.
+				// We pre-merge that data prior to runner mapper.merge.
+				if mapper.newIamUpdater != nil && !c.offline && existing.IAMPolicy == nil {
+					updater, err := mapper.newIamUpdater(data, c.cfg)
+					if err != nil {
+						return err
+					}
+					iamPolicy, err := updater.GetResourceIamPolicy()
+					if err != nil {
+						return err
+					}
+					fmt.Printf("%+v\n", iamPolicy)
+					var bindings []converter.IAMBinding
+					for _, b := range iamPolicy.Bindings {
+						bindings = append(bindings, converter.IAMBinding{
+							Role:    b.Role,
+							Members: b.Members,
+						})
+					}
+					converterAsset.IAMPolicy = &converter.IAMPolicy{
+						Bindings: bindings,
+					}
+				}
+
+				converted = mapper.merge(converterAsset, converted)
```

Comment:
Does this handle bindings in a TF config that will remove a similar binding on the IAM policy?

Reply:
hm, yeah, there's a good chance the merging logic needs to be refactored to account for deleting assets.

Reply:
Some additional rambling on this point:

Historically, terraform-validator didn't need to care about the diff between the current vs planned assets, because it just needed to make sure the planned state was valid. So if something wasn't present in planned assets, it wouldn't get validated, and that was probably fine, because it would be deleted.

So, the bug we would be introducing by not properly handling deleted iam assets is that validation could get tripped up with something being invalid that has actually already been removed in the plan. For simulator (the root feature request) they would never be able to tell when something had been removed - maybe even worse than the situation they're in now.

Reply:
I don't think there's a way to resolve this without making terraform-validator diff-aware - if we only look at the planned resources there's no way to know what Bindings need to be deleted, for example. There is existing work from @cledoux to add a ReadCurrentAssets function in order to be able to diff; however, it functions by reading the current state and planned state separately by diffing them, which wouldn't allow us to access the diff data while adding a single resource (so it doesn't help us handle fine-grained resources.)

I think it might be worthwhile to refactor the system to work with resource changes instead so that each resource being processed has access to both the previous and current state and can understand what changes have occurred.